### PR TITLE
Create PROTOCOL.md

### PR DESCRIPTION
**Pull request by mocax**
Draft the wire protocol spec for Twirp based on latest discussions. It reflects the existing Twirp wire spec, plus best practice from proto3, gRPC, HTTP, and JSON. The spec should be usable for developers to build clients and servers using Twirp wire protocol without much dependency.

*Issue #, if available:*
#71  
*Description of changes:*
Write up the Twirp wire protocol.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by daroot** [comment]
`application/x-protobuf` or `application/protobuf`?   The code uses the latter, not the former.

**Comment by slaskis** [comment]
Charles Proxy [seems to support](https://www.charlesproxy.com/documentation/using-charles/protocol-buffers/) either `application/x-protobuf` or `application/x-google-protobuf` and it's nice to have supported tooling. Although maybe using `application/protobuf` in Twirp might coax the Charles devs to add that Content-Type too...

**Comment by mocax** [comment]
Google didn't complete its IETF application, https://tools.ietf.org/html/draft-rfernando-protocol-buffers-00. So technically it still is `application/x-protobuf`.

File: PROTOCOL.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+To use Twirp, developers first define their APIs using proto files,
+then use Twirp tools to generate the client and the server libraries.
+The generated libraries implement the Twirp wire protocol, using the
+standard HTTP library provided by the programming language runtime or
+the operating system. Once the client and the server are implemented,
+the client can communicate with the server by making RPC calls.
+
+The Twirp wire protocol supports both binary and JSON encodings of
+proto messages, and works with any HTTP client and any HTTP version.
+However, certain capabilities may be limited by the actual HTTP
```
Comment (spenczar) [inline PROTOCOL.md]: I don't think it's worth including this caveat until we can actually say this precisely.

File: PROTOCOL.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+library being used.
+
+### URLs
+
+**URL ::= Base-URL "/" [ Package "." ] Interface "/" Method**
```
Comment (spenczar) [inline PROTOCOL.md]: Please change `Interface` to `Service`.

Also, the current routing is `URL ::= Base-URL "/twirp/" [ Package "." ] Service "/" Method`, please update this to match the `/twirp` prefix.

File: PROTOCOL.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+The Twirp wire protocol uses HTTP URLs to directly specify the RPC
```
Comment (spenczar) [inline PROTOCOL.md]: remove "directly"

File: PROTOCOL.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+endpoints on the server for sending the requests. such direct mapping
```
Comment (spenczar) [inline PROTOCOL.md]: Capitalize "direct"

File: PROTOCOL.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+makes the request routing simple and efficient. The Twirp URLs have
+the following components.
+
+* **Base-URL** is the virtual location of a Twirp API server, which is
+  typically published via API documentation or service discovery.  For
+  example, "https://example.com/apis".
```
Comment (spenczar) [inline PROTOCOL.md]: This is a bad example right now, as the base URL in v5 cannot safely include a path component.

I think we should be more specific here. The Base URL should contain only the "scheme" and "authority" (commonly known as "host and port") portions of a [RFC 3986 URI](https://tools.ietf.org/html/rfc3986).

We will likely change this to allow a path component in v6, as discussed in #55.

File: PROTOCOL.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
+# Twirp Wire Protocol
+
+This document defines the Twirp wire protocol over HTTP.
```
Comment (spenczar) [inline PROTOCOL.md]: Can we include somewhere that this describes version 5 of Twirp?

File: PROTOCOL.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+* **Interface** is the proto `service` name for an API. For example,
+  `CalendarService`.
+
+* **Method** is the proto `rpc` name for an API method. For example,
+  `CreateEvent`.
+
+### Requests
+
+**Request ::= Request-Headers Request-Body**
```
Comment (spenczar) [inline PROTOCOL.md]: I'm not sure what this line is getting us. Can we delete it?

File: PROTOCOL.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+Twirp always uses HTTP POST method to send requests, because it
+closely matches the semantics of RPC methods.
+
+The **Request-Headers** are normal HTTP headers. The Twirp wire
+protocol uses the following headers.
+
+* **Authorization** header is often used to pass user credentials
```
Comment (spenczar) [inline PROTOCOL.md]: This is unused by Twirp, please remove.

File: PROTOCOL.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  from the client to the server, such as OAuth access token or
+  JWT token.
+
+* **Content-Type** header indicates the proto message encoding, which
+  should be one of "application/x-protobuf", "application/json". The
```
Comment (spenczar) [inline PROTOCOL.md]: We accept `application/protobuf`, not `application/x-protobuf`.

If we change that in v6, we can say so in a v6 spec, but we should describe the current state of things in this PR.

File: PROTOCOL.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+  server uses this value to decide how to parse the request body,
+  and encode the response body.
+
+* **User-Agent** header indicates the client application and its
```
Comment (spenczar) [inline PROTOCOL.md]: This is unused by Twirp, please remove.

File: PROTOCOL.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+  runtime environment. While the server should not use this
+  information for request processing, this header is heavily used
+  for analytics and troubleshooting purposes.
+
+* **RPC-Timeout** header indicates the client-specified request
```
Comment (spenczar) [inline PROTOCOL.md]: This is unused by Twirp, please remove.

File: PROTOCOL.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+  server should use a pre-configured timeout value, by default it
+  should be 10 seconds.
+
+The **Request-Body** is the encoded request message, contained in the
+HTTP request body. The encoding is specified by the `Content-Type`
+header.
+
+### Responses
+
+**Response ::= Response-Headers Response-Body**
```
Comment (spenczar) [inline PROTOCOL.md]: I think we can drop this line, too.

File: PROTOCOL.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+The **Response-Headers** are just normal HTTP response headers. The
+Twirp wire protocol uses the following headers.
+
+* **Content-Type** The value should be either "application/x-protobuf"
```
Comment (spenczar) [inline PROTOCOL.md]: As above, should be `application/protobuf`

File: PROTOCOL.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+## Example
+
+The following example shows a simple Echo API definition and its
+corresponding wire payloads.
+
+The example assumes the server base URL is "https://example.com".
+
+```proto
+package twirp;
```
Comment (spenczar) [inline PROTOCOL.md]: Missing `syntax = "proto3";` at the top of this proto file.

File: PROTOCOL.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+  string message;
+}
+```
+
+**Proto Request**
+
+```
+POST /twirp.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/x-protobuf
```
Comment (spenczar) [inline PROTOCOL.md]: should be `application/protobuf`

File: PROTOCOL.md [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+Content-Length: 15
+
+<encoded HelloRequest>
```
Comment (spenczar) [inline PROTOCOL.md]: Should we try to represent the encoded protobuf here in some way? I think that might be valuable.

File: PROTOCOL.md [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+The following example shows a simple Echo API definition and its
+corresponding wire payloads.
+
+The example assumes the server base URL is "https://example.com".
+
+```proto
+package twirp;
+
+service Echo {
```
Comment (spenczar) [inline PROTOCOL.md]: Let's change this to the classic Twirp example, [`Haberdasher`](https://github.com/twitchtv/twirp/blob/master/example/service.proto) :)

File: PROTOCOL.md [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+
+## Errors
+
+If an error occurs when the server processes a request, the server
+must return an error payload as the response message, and correctly
+set the HTTP status code. Please see
+[`google.rpc.Code`](https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)
+on how to map typical server errors to HTTP status codes.
+
+### Timeout errors
```
Comment (spenczar) [inline PROTOCOL.md]: This is not true of Twirp, please remove this section.

File: PROTOCOL.md [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+For a single request, there is a client-specified timeout and a
+server-configured timeout. If a request misses the server-configured
+timeout, the server must return a `503` error. If a request misses
+the client-specified timeout that it is shorter then the server-
+configured timeout, the server must return a `504` error.
+This allows more accurate measurement of the server availability.
+
+### Network errors
+
+If a client fails to reach the server due to network errors, the
```
Comment (spenczar) [inline PROTOCOL.md]: I don't think we should specify this. I don't think client libraries should be exposing HTTP status codes, just Twirp error codes.

File: PROTOCOL.md [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+
+{"message":"Hello, World!"}
+```
+
+## Errors
+
+If an error occurs when the server processes a request, the server
+must return an error payload as the response message, and correctly
+set the HTTP status code. Please see
+[`google.rpc.Code`](https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)
```
Comment (spenczar) [inline PROTOCOL.md]: We don't strictly follow `google.rpc.Code`.  We use our own error codes, heavily based on those but with some modifications: the codes are a string on the wire for human readability, and we add the `bad_route` error code. See https://github.com/twitchtv/twirp/blob/master/errors.go#L138-L275.

File: PROTOCOL.md [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+## Errors
+
+If an error occurs when the server processes a request, the server
+must return an error payload as the response message, and correctly
```
Comment (spenczar) [inline PROTOCOL.md]: I think we should describe Twirp's error payload in detail here.

Twirp error responses are always JSON-encoded (*regardless* of the request's Content-Type), with a corresponding `Content-Type: application/json` header. This ensures that they are human-readable in any setting.

They are a JSON object with three keys:

 - **code**: One of the Twirp error codes as a string.
 - **msg**: A human-readable message describing the error as a string.
 - **meta**: An object with string keys and values holding arbitrary additional metadata describing the error.

For example:
```json
{
    "code": "permission_denied",
    "msg": "thou shall not pass",
    "meta": {
        "target": "Balrog"
    }
}
```

See https://github.com/twitchtv/twirp/wiki/Errors.

File: PROTOCOL.md [hunk 22]
```diff
@@ -0,0 +1,193 @@
+# Twirp Wire Protocol
+
+This document defines the Twirp wire protocol over HTTP.
+
+## Conventions
+
+The requirement level keywords "MUST", "MUST NOT", "REQUIRED",
```
Comment (spenczar) [inline PROTOCOL.md]: None of these keywords are used in this document. Let's drop this paragraph.

File: PROTOCOL.md [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+"SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY",
+and "OPTIONAL" used in this document are to be interpreted as
+described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+The grammar rules used in this document are using [ABNF
```
Comment (spenczar) [inline PROTOCOL.md]: We only use ABNF in one place where it's useful, which is in describing the URL. Let's delete this paragraph and just say something like `In [ABNF syntax](https://tools.ietf.org/html/rfc5234), Twirp's URLs follow this format: ...`

Then, we can get rid of this `Conventions` section altogether, which is good.

File: PROTOCOL.md (hunk 17 above)
Comment (mocax) [reply PROTOCOL.md]: Echo is the most classic example for network APIs used in many places. As a protocol spec, we should use the example most developers are already familiar with.

For programming languages, everyone chooses hello-world. I think we should just follow the historical tradition.

File: PROTOCOL.md (hunk 16 above)
Comment (mocax) [reply PROTOCOL.md]: This is a public spec. I believe 99% developers don't know or care about proto encoding format. Let's wait and see. We can always add it if many people ask for it.

**Comment by mocax** [comment]
Addressed most comments.

File: PROTOCOL.md [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+
+message HelloResponse {
+  string message;
+}
+```
+
+**Proto Request**
+
+```
+POST /twirp.Echo/Hello HTTP/1.1
```
Comment (spenczar) [inline PROTOCOL.md]: The URL here should be `/twirp/twirp.Echo/Hello`. Same for JSON.

The stuttering of 'twirp' here is not ideal. Could you change the package to something like `package example.echoer`?

**Comment by mocax** [comment]
Done.

**Pull request merged**